### ISSUE TITLE
Class registry doesn't respect namespaces

### ISSUE THREAD

**Issue by jcrist** [sympy/sympy#7893]
In SymPy, the class registry `C` is used internally to prevent circular imports in the code. All classes that are instances of `Basic` are added to this registry via a metaclass. The issue here is that the registry doesn't respect namespaces. Class names that are the same will result in the registry being changed upon class initialization. 

In the following example, I create a user created `Function` class subclassing basic, and show how it changes the internal class registry in sympy:

```
>>> import sympy as sym
>>> sym.C.Function is sym.Function
True
>>> class Function(sym.Basic):
   ...:     pass 
>>> f = Function()
>>> sym.C.Function is sym.Function
False
>>> sym.C.Function is Function
True
```

This has already been shown to cause failures, as shown in PR #7608, where another `Function` class is created to be part of the FOL module. During the doctests, creation of a FOL `Function` overrides the normal core `Function`, which causes errors in the `fcode` tests as `C.Function` is no longer a core `Function`, but a FOL `Function`. The only reason this was caught was due to the order of tests, so we got lucky.

---

Because the registry is necessary, I don't think we can remove it. Instead I vote that it be modified to respect the namespaces, so there could be a `C.FOL.Function` (or whatever the path actually is). I'm not sure how this would work though. A solution needs to be found though, as this prevents any sympy objects ever having the same name, even if they live in different namespaces or are user created.

**Comment by toolforger** [comment]
Can we close this? It is being solved through #7124 .

**Comment by oscarbenjamin** [comment]
The class registry is now deprecated so I think this should probably be close.
